src/ray/object_manager/plasma/plasma_allocator.cc

```diff
@@ -31,68 +31,104 @@ void dlfree(void *mem);
+const size_t kAllocationAlignment = 64;
 
-void *PlasmaAllocator::Memalign(size_t alignment, size_t bytes) {
+absl::optional<Allocation> BuildAllocation(void *addr, size_t size) {
+  if (addr == nullptr) {
+    return absl::nullopt;
+  }
+  MEMFD_TYPE fd;
+  int64_t mmap_size;
+  ptrdiff_t offset;
+
+  if (GetMallocMapinfo(addr, &fd, &mmap_size, &offset)) {
+    return Allocation(addr, static_cast<int64_t>(size), std::move(fd), offset,
+                      0 /* device_number*/, mmap_size);
+  }
+  return absl::nullopt;
+}
+
+}  // namespace
+
+/* static */ PlasmaAllocator &PlasmaAllocator::GetInstance() {
+  static PlasmaAllocator instance(kAllocationAlignment);
+  return instance;
+}
+
+PlasmaAllocator::PlasmaAllocator(size_t alignment)
+    : kAlignment(alignment), allocated_(0), fallback_allocated_(0), footprint_limit_(0) {}
+
+absl::optional<Allocation> PlasmaAllocator::Allocate(size_t bytes) {
   if (!RayConfig::instance().plasma_unlimited()) {
     // We only check against the footprint limit in limited allocation mode.
     // In limited mode: the check is done here; dlmemalign never returns nullptr.
     // In unlimited mode: dlmemalign returns nullptr once the initial /dev/shm block
     // fills.
     if (allocated_ + static_cast<int64_t>(bytes) > footprint_limit_) {
-      return nullptr;
+      return absl::nullopt;
     }
   }
   RAY_LOG(DEBUG) << "allocating " << bytes;
-  void *mem = dlmemalign(alignment, bytes);
+  void *mem = dlmemalign(kAlignment, bytes);
   RAY_LOG(DEBUG) << "allocated " << bytes << " at " << mem;
   if (!mem) {
-    return nullptr;
+    return absl::nullopt;
   }
   allocated_ += bytes;
-  return mem;
+  return BuildAllocation(mem, bytes);
 }
 
-void *PlasmaAllocator::DiskMemalignUnlimited(size_t alignment, size_t bytes) {
+absl::optional<Allocation> PlasmaAllocator::FallbackAllocate(size_t bytes) {
   // Forces allocation as a separate file.
   RAY_CHECK(dlmallopt(M_MMAP_THRESHOLD, 0));
-  void *mem = dlmemalign(alignment, bytes);
+  RAY_LOG(DEBUG) << "fallback allocating " << bytes;
+  void *mem = dlmemalign(kAlignment, bytes);
+  RAY_LOG(DEBUG) << "allocated " << bytes << " at " << mem;
   // Reset to the default value.
   RAY_CHECK(dlmallopt(M_MMAP_THRESHOLD, MAX_SIZE_T));
+
   if (!mem) {
-    return nullptr;
+    return absl::nullopt;
   }
+
   allocated_ += bytes;
   // The allocation was servicable using the initial region, no need to fallback.
   if (IsOutsideInitialAllocation(mem)) {
     fallback_allocated_ += bytes;
   }
-  return mem;
+  return BuildAllocation(mem, bytes);
 }
 
-void PlasmaAllocator::Free(void *mem, size_t bytes) {
-  RAY_LOG(DEBUG) << "deallocating " << bytes << " at " << mem;
-  dlfree(mem);
-  allocated_ -= bytes;
-  if (RayConfig::instance().plasma_unlimited() && IsOutsideInitialAllocation(mem)) {
-    fallback_allocated_ -= bytes;
+void PlasmaAllocator::Free(const Allocation &allocation) {
+  RAY_CHECK(allocation.address != nullptr);
```

Comment:
Great! Can you also add a comment here btw? 
```suggestion
  RAY_CHECK(allocation.address != nullptr) << "Cannot free the nullptr"; 
```